app.go

```diff
@@ -475,6 +506,39 @@ func New(opts ...Option) *App {
 		app.provide(p)
 	}
 
+	// If WithLogger and Printer are both provided, WithLogger takes precedence.
+	// If user supplied an fxlog constructor, we need to have dig's container
+	// initialize it and grab all of arguments from the graph and then return.
+	if app.fxlogConstructor != nil {
```

Comment:
I'm not sure but I think this check should be after `opt.apply` but before `app.provide`? By this point in time, we've already provided a bunch of things to the container and logged about them. We also probably need an `else` clause to use the default logger when this is unspecified?

So what we need here is probably something like:

```
for _, opt := range opts {
  opt.apply(app)
}

var connectLogger interface{}
if app.fxLogConstructor != nil {
    // There's a user provided constructor. Log to buffer until we connect.
    app.log = app.logBuffer
    connectLogger = func(log fxlog.Logger) {
        app.logBuffer.Connect(log)
    }
} else {
    app.log = // put the default logger here
}

// ... app.provide(p)

if connectLogger != nil {
    app.container.Invoke(connectLogger)
}
```

The net result of this will be:

- Existing logging code remains unchanged and always logs to app.log
- app.log is a buffer if the user provided a logger. Otherwise it's not.
- If the user provided a logger, then the first thing we do after all provides is connect the logger

---

On top of this, we can actually get away with not having logBuffer as a field on app:

```
if app.fxLogConstructor != nil {
  var buffer logBuffer
  app.log = &buffer
  connectLogger = buffer.Connect // func(fxlog.Logger)
}
```

I slightly prefer this so that we don't accidentally use the logBuffer directly ever.

---

Finally, and this is optional, but the the Invoke can instead be this:

```
func(log fxlog.Logger) {
  buffer.FlushTo(log)
  app.log = log  // replace buffer with real logger
}
```

This will remove the buffer completely from the equation once things are ready and remove the if-else from the buffer. The buffer would just record things in memory until the replacement is available at which point we flush to replacement and replace.

Reply:
_I'm not sure but I think this check should be after opt.apply but before app.provide_

So WithLogger is going to be part of Provide, so I think we need this after app.provide as `fxlogConstructor` is populated by WithLogger opt.apply but it'll need `zap.Logger` initialized by something like `zapfx`.

Good call on trying to simplify logBuffer, let me see how I can make it work.

Reply:
If we do this after the `app.provides`, then we've already called `container.Provide` and logged a bunch before we've even considered whether there was a user-provided logger, right?

It's an option, but it's a "more important" option than provides. Like, we run Provides first, and Invokes later. We should examine WithLogger *before* we start logging things the same way we consider the current fx.Printer before anything else.

To be clear: we won't construct the logger until after app.provides, but we need to know whether we're buffering or using the default logger before we do app.provides. Please read the code samples above again.